### PR TITLE
Fix longer type-only property access in non-emitting heritage clauses

### PR DESCRIPTION
**Pull request by andrewbranch**
Reported in of https://github.com/microsoft/TypeScript/issues/36478#issuecomment-595879205 / #36496

Would like to get this in 3.8.4 if possible.

### CODE REVIEW

File: src/compiler/utilities.ts [hunk 1]
```diff
@@ -6292,8 +6292,18 @@ namespace ts {
 
     /** Returns true for the first identifier of 1) an `implements` clause, and 2) an `extends` clause of an interface. */
     function isFirstIdentifierOfNonEmittingHeritageClause(node: Node): boolean {
-        // Number of parents to climb from identifier is 2 for `implements I`, 3 for `implements x.I`
```
Comment (andrewbranch) [inline src/compiler/utilities.ts]: At the time I was thinking that the left side of a property access was the _top_ of the tree 🤦‍♂

File: src/compiler/utilities.ts [hunk 2]
```diff
@@ -6292,8 +6292,18 @@ namespace ts {
 
     /** Returns true for the first identifier of 1) an `implements` clause, and 2) an `extends` clause of an interface. */
     function isFirstIdentifierOfNonEmittingHeritageClause(node: Node): boolean {
```
Comment (DanielRosenwasser) [inline src/compiler/utilities.ts]: I don't get where "first"-ness plays into the implementation here, or why you'd want it. If you're only looking the first part of an `extends` or `implements`, does this mean that the following

```ts
import type { Baz } from "whatever

interface Bar {}

interface Foo extends Bar, Baz {}
```

still doesn't work?

Comment (andrewbranch) [reply src/compiler/utilities.ts]: It’s not the first-ness of the heritage clause, it’s the first-ness of the identifier that we’re looking at. Which also doesn’t strictly matter, but this function will never run on something in the middle/end of a property access expression, so originally it was intended to be a performance enhancement to bail early on those. But, that was, again, written with an inverted tree in mind (🤦‍♂ 🤦‍♂ 🤦‍♂), so now this is in fact a misnomer, though the logic is good. I’ll rename it.

**Comment by andrewbranch** [comment]
@typescript-bot cherry-pick this to release-3.8

**Pull request merged**